internal/state/store.go

```diff
@@ -3,83 +3,280 @@ package state
-	if s.gc != nil && s.gc.Generation == gc.Generation {
-		resourceChanged = false
+	oldObj, exist := s.store.get(obj, client.ObjectKeyFromObject(obj))
+	s.store.upsert(obj)
+
+	if !s.trackedUpsertDeleteGKVs.contains(s.extractGVK(obj)) {
+		return false
 	}
 
-	s.gc = gc
+	return !exist || obj.GetGeneration() != oldObj.GetGeneration()
+}
+
+func (s *changeTrackingUpdater) Upsert(obj client.Object) {
+	s.assertSupportedGVK(s.extractGVK(obj))
 
-	s.changed = s.changed || resourceChanged
+	changingUpsert := s.upsert(obj)
+	s.capturer.Capture(obj)
+
+	s.changed = s.changed || changingUpsert || s.capturer.Exists(obj, client.ObjectKeyFromObject(obj))
 }
 
-func (s *store) captureGatewayChange(gw *v1beta1.Gateway) {
-	resourceChanged := true
+func (s *changeTrackingUpdater) delete(objType client.Object, nsname types.NamespacedName) (changed bool) {
+	objTypeGVK := s.extractGVK(objType)
+
+	if !s.persistedGKVs.contains(objTypeGVK) {
+		return false
+	}
 
-	// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-	prev, exist := s.gateways[client.ObjectKeyFromObject(gw)]
-	if exist && gw.Generation == prev.Generation {
-		resourceChanged = false
+	_, exist := s.store.get(objType, nsname)
+	if !exist {
+		return false
 	}
+	s.store.delete(objType, nsname)
+
+	return s.trackedUpsertDeleteGKVs.contains(objTypeGVK)
+}
 
-	s.gateways[client.ObjectKeyFromObject(gw)] = gw
+func (s *changeTrackingUpdater) Delete(objType client.Object, nsname types.NamespacedName) {
+	s.assertSupportedGVK(s.extractGVK(objType))
+
+	changingDelete := s.delete(objType, nsname)
+
+	s.changed = s.changed || changingDelete || s.capturer.Exists(objType, nsname)
+
+	s.capturer.Remove(objType, nsname)
+}
 
-	s.changed = s.changed || resourceChanged
+// getAndResetChangedStatus returns true if the previous updates (Upserts/Deletes) require an update of
+// the configuration of the data plane. It also resets the changed status to false.
+func (s *changeTrackingUpdater) getAndResetChangedStatus() bool {
+	changed := s.changed
+	s.changed = false
+	return changed
 }
 
-func (s *store) captureHTTPRouteChange(hr *v1beta1.HTTPRoute) {
-	resourceChanged := true
-	// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-	prev, exist := s.httpRoutes[client.ObjectKeyFromObject(hr)]
-	if exist && hr.Generation == prev.Generation {
-		resourceChanged = false
+type upsertValidatorFunc func(obj client.Object) error
+
+// validatingUpsertUpdater is an Updater that validates an object before upserting it.
+// If the validation fails, it deletes the object and records an event with the validation error.
+type validatingUpsertUpdater struct {
+	updater          Updater
+	eventRecorder    record.EventRecorder
+	upsertValidators []upsertValidatorFunc
+}
+
+func newValidatingUpsertUpdater(
+	updater Updater,
+	eventRecorder record.EventRecorder,
+	upsertValidators []upsertValidatorFunc,
```

Comment:
Why a list of upsertValidatorFunc?

Reply:
Seems like a premature extension point. Do we have immediate plans to add another validator?

Reply:
fair enough